go/adbc/driver/bigquery/statement.go

```diff
@@ -0,0 +1,525 @@
+// For queries expected to be executed repeatedly, Prepare should be
+// called before execution.
+func (st *statement) SetSqlQuery(query string) error {
+	st.query.QueryConfig.Q = query
+	return nil
+}
+
+// ExecuteQuery executes the current query or prepared statement
+// and returns a RecordReader for the results along with the number
+// of rows affected if known, otherwise it will be -1.
+//
+// This invalidates any prior result sets on this statement.
+func (st *statement) ExecuteQuery(ctx context.Context) (array.RecordReader, int64, error) {
+	parameters, err := st.getQueryParameter()
+	if err != nil {
+		return nil, 0, err
+	}
+	if parameters != nil {
+		st.query.QueryConfig.Parameters = parameters
+	}
+	reader, affectedRows, err := newRecordReader(ctx, st.query, st.connectionImpl.Alloc)
+	if err != nil {
+		return nil, -1, err
+	}
+	return reader, affectedRows, nil
+}
+
+// ExecuteUpdate executes a statement that does not generate a result
+// set. It returns the number of rows affected if known, otherwise -1.
+func (st *statement) ExecuteUpdate(ctx context.Context) (int64, error) {
+	return -1, adbc.Error{
+		Code: adbc.StatusNotImplemented,
+		Msg:  "ExecuteUpdate not yet implemented for BigQuery driver",
+	}
+}
+
+// ExecuteSchema gets the schema of the result set of a query without executing it.
+func (st *statement) ExecuteSchema(ctx context.Context) (*arrow.Schema, error) {
+	return nil, adbc.Error{
+		Code: adbc.StatusNotImplemented,
+		Msg:  "ExecuteSchema not yet implemented for BigQuery driver",
+	}
+}
+
+// Prepare turns this statement into a prepared statement to be executed
+// multiple times. This invalidates any prior result sets.
+func (st *statement) Prepare(_ context.Context) error {
+	return nil
+}
+
+// SetSubstraitPlan allows setting a serialized Substrait execution
+// plan into the query or for querying Substrait-related metadata.
+//
+// Drivers are not required to support both SQL and Substrait semantics.
+// If they do, it may be via converting between representations internally.
+//
+// Like SetSqlQuery, after this is called the query can be executed
+// using any of the Execute methods. If the query is expected to be
+// executed repeatedly, Prepare should be called first on the statement.
+func (st *statement) SetSubstraitPlan(plan []byte) error {
+	return adbc.Error{
+		Code: adbc.StatusNotImplemented,
+		Msg:  "Substrait not yet implemented for BigQuery driver",
+	}
+}
+
+func arrowValueToQueryParameterValue(value arrow.Array) bigquery.QueryParameter {
+	parameter := bigquery.QueryParameter{}
+	if value.IsNull(0) {
+		parameter.Value = &bigquery.QueryParameterValue{
+			Type: bigquery.StandardSQLDataType{
+				TypeKind: "NULL",
+			},
+			Value: "NULL",
+		}
+		return parameter
+	}
+
+	// https://cloud.google.com/bigquery/docs/reference/storage#arrow_schema_details
+	// https://cloud.google.com/bigquery/docs/reference/rest/v2/StandardSqlDataType#typekind
+	switch value.DataType().ID() {
```

Comment:
what about list and struct types? If we aren't going to implement them yet, we need to error on them please. Same for any other types (Map, Union, etc.)

Reply:
We should also either implement for dictionary types or add a todo and return an error for them

Reply:
Got it! I'll leave a todo here and implement them in separate PRs.